pkg/migration/manager.go

```diff
@@ -0,0 +1,283 @@
+) *Manager {
+	return &Manager{
+		dialer:   dialer,
+		executor: executor,
+		nl:       nl,
+		db:       db,
+	}
+}
+
+// IterateRangeDescriptors is a primitive for constructing migrations. It
+// provides a handle on every range in descriptor in the system, which callers
+// can then use to send out arbitrary KV requests to in order to run arbitrary
+// KV-level migrations. These requests will typically just be the Migrate
+// request, with code added within pkg/kv to do the specific things intended for
+// the specified version.
+//
+// TODO(irfansharif): We want to eventually be able to send out a Migrate
+// request spanning the entire keyspace. We'll need to make sure all stores have
+// synced once to persist any raft command applications.
+//lint:ignore U1001 unused
+func (h *Helper) IterateRangeDescriptors(
+	ctx context.Context, blockSize int, f func(...roachpb.RangeDescriptor) error,
+) error {
+	descriptors := make([]roachpb.RangeDescriptor, blockSize)
+	if err := h.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		// Iterate through the meta ranges to pull out all the range descriptors.
+		return txn.Iterate(ctx, keys.MetaMin, keys.MetaMax, blockSize,
+			func(rows []kv.KeyValue) error {
+				for i, row := range rows {
+					if err := row.ValueProto(&descriptors[i]); err != nil {
+						return errors.Wrapf(err, "%s: unable to unmarshal range descriptor", row.Key)
+					}
+				}
+
+				// Invoke f with the current chunk (of size ~blockSize) of range
+				// descriptors.
+				if err := f(descriptors...); err != nil {
+					return err
+				}
+
+				return nil
+			})
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RequiredNodes returns the node IDs for all nodes that are currently part of
+// the cluster (i.e. they haven't been decommissioned away). Migrations have the
+// pre-requisite that all required nodes are up and running so that we're able
+// to execute all relevant node-level operations on them. If any of the nodes
+// are found to be unavailable, an error is returned.
+func (h *Helper) RequiredNodes(ctx context.Context) ([]roachpb.NodeID, error) {
+	var nodeIDs []roachpb.NodeID
+	livenesses, err := h.nl.GetLivenessesFromKV(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, liveness := range livenesses {
+		if liveness.Membership.Decommissioned() {
+			continue
+		}
+
+		live, err := h.nl.IsLive(liveness.NodeID)
+		if err != nil {
+			return nil, err
+		}
+		if !live {
+			return nil, errors.Newf("n%d required, but unavailable", liveness.NodeID)
+		}
+		nodeIDs = append(nodeIDs, liveness.NodeID)
+	}
+
+	return nodeIDs, nil
+}
+
+// Log lets migrations insert arbitrary log events. These events are recorded in
+// a system table, and are available for introspection into the progress of the
+// specific ongoing migration.
+//
+// TODO(irfansharif): Implement this.
+//lint:ignore U1001 unused
+func (h *Helper) Log(event string) error {
```

Comment:
Let's strip everything that isn't essential for now.

Reply:
A version of this is now implemented in #57445.